feat: Managed Events V2

alishaz-polymath opened the pull request:
## What does this PR do?

Continues the work done by @leog in #11958 👏 

Fixes #9307 #11941 

### Screenshots

<kbd><img src="https://github.com/calcom/cal.com/assets/467258/28379efe-18e6-4782-a0b7-63ccda794965" /></kbd>

<kbd><img src="https://github.com/calcom/cal.com/assets/467258/657e8a63-3519-4526-af55-b86bf9bdb889" /></kbd>

<kbd><img src="https://github.com/calcom/cal.com/assets/467258/4e9c4149-7a9c-42a0-b600-08c3631aeac4" /></kbd>

<kbd><img src="https://github.com/calcom/cal.com/assets/467258/c9edc1f6-ae58-4667-b4f9-d3b3a37b531d" /></kbd>

[Loom](https://www.loom.com/share/483902d4c4154eea97dbbf2688feac3a)

- [x] This PR allows the Managed Events to be configurable enough that the Admin can lock/unlock fields. 
- [x] The values of the unlocked fields can be modified by the members. 
- [x] The field values updated by members persist if the Admin makes changes to any other fields in the event type, which is the desired outcome

Current quirks remaining and being fixed (checked once fixed)

- [x] The locked/unlocked changed state not persisting when moving from one tab to another within the event type as an Admin
- [x] Private URLs are currently excluded from being locked/unlocked and have thus been put as disabled. Currently working to sort this out
- [x] Fields contains "unlocked" labels in non-managed events as well, so they need to be removed to avoid any confusion

## Requirement/Documentation

<!-- Please provide all documents that are important to understand the reason of that PR. -->

- [REF](https://www.figma.com/file/UajFu4M1APhn2AywAEJkJr/New-Features?type=design&node-id=8320-40915)
- For the locked/unlocked UI, went ahead with the badge with title and the switch as it was the one that can be universally applied to any type of field, whether is a text or a setting toggle.


## Type of change

<!-- Please delete bullets that are not relevant. -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Mandatory Tasks

- [ ] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

<!-- Remove bullet points below that don't apply to you -->

alishaz-polymath committed:
init

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
Fix event limit interlinked switches

alishaz-polymath committed:
wip

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
WIP

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
fix bugs

alishaz-polymath committed:
lock hashed link until further notice

alishaz-polymath committed:
lock private url for managed type to stop confusion

alishaz-polymath committed:
fix recurring

alishaz-polymath committed:
prettier fix

CarinaWolli reviewed (commented):
Nice work, looks pretty good already 👏🏻 

Some things I found during testing: 

- When locking/unlocking fields and switching tabs before saving the changes are lost
- Any change in the parent resets everything from the children to the default values of the parent
  - Not sure about the actual field that is changed if it is unlocked ->  If an unlocked field is changed in the parent, should this change the value of all children?  
- When locking workflows and a user already has a personal workflow assigned (as it was unlocked before), it stays active. For all other field we change back to the default settings if locked, so I think we should also disable all personal workflows once locked

CarinaWolli commented on packages/features/eventtypes/lib/types.ts:
```diff
@@ -0,0 +1,77 @@
+  isManaged?: boolean;
+};
+
+export type FormValues = {
```
misses the following fields: 

```
isInstantEvent: boolean; 
lockTimeZoneToggleOnBookingPage: boolean;
onlyShowFirstAvailableSlot: boolean;
```

CarinaWolli commented on apps/web/pages/event-types/[type]/index.tsx:
```diff
@@ -84,68 +77,6 @@ const EventWebhooksTab = dynamic(() =>
 
 const ManagedEventTypeDialog = dynamic(() => import("@components/eventtype/ManagedEventDialog"));
 
-export type FormValues = {
```
import for `FormValues` needs to be fixed in `InstantEventController.tsx`

alishaz-polymath commented:
> Nice work, looks pretty good already 👏🏻
> 
> Some things I found during testing:
> 
>     * When locking/unlocking fields and switching tabs before saving the changes are lost

Thank you for this, will take a look 🙏 

> 
>     * Any change in the parent resets everything from the children to the default values of the parent
>       
>       * Not sure about the actual field that is changed if it is unlocked ->  If an unlocked field is changed in the parent, should this change the value of all children?

I think it should. If the Admin/Parent event type field is changed, it should apply to the children, locked or unlocked. Although, I see the problem where the Admin wishes to change the default (for future users) but that might just be over-engineering 🤔 

> 
>     * When locking workflows and a user already has a personal workflow assigned (as it was unlocked before), it stays active. For all other field we change back to the default settings if locked, so I think we should also disable all personal workflows once locked

Ah, right. I had a similar issue with Private URL as well. I'll check if the fix is straightforward for this. I have currently locked private url, for example. If the fix seems like a more complex thing to fix, probably better to do a follow up and fix exactly this kind of issue in v2.1. I'll check and report back 🙏 

Thank you for these shouts @CarinaWolli 👏

alishaz-polymath committed:
early review fixes ...

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

CarinaWolli commented:
> I think it should. If the Admin/Parent event type field is changed, it should apply to the children, locked or unlocked. Although, I see the problem where the Admin wishes to change the default (for future users) but that might just be over-engineering 🤔

I think the actual change is fine to be applied in all children. We can revisit if users run into problems or would expect different 👍

> Any change in the parent resets everything from the children to the default values of the parent 

This definitely shouldn't happen, only the field changed should be affected

alishaz-polymath commented:
> > I think it should. If the Admin/Parent event type field is changed, it should apply to the children, locked or unlocked. Although, I see the problem where the Admin wishes to change the default (for future users) but that might just be over-engineering 🤔
> 
> I think the actual change is fine to be applied in all children. We can revisit if users run into problems or would expect different 👍
> 
> > Any change in the parent resets everything from the children to the default values of the parent
> 
> This definitely shouldn't happen, only the field changed should be affected

Oh right, yeah this is what I meant. Any specific field change should update the value of that field in child event types

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
--WIP to send only changed form fields in updateMutation

alishaz-polymath committed:
WIP with some type fixes

alishaz-polymath committed:
Revert "WIP with some type fixes"

This reverts commit 00f10b772d6d08af17e7c9bd2d3601d93035d9b3.

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
post merge conflict resolution fixes

alishaz-polymath committed:
further type fix

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
fixed test

leog commented:
@alishaz-polymath is there anything I can help with? Sorry I didn't catch this up before.

alishaz-polymath commented:
> @alishaz-polymath is there anything I can help with? Sorry I didn't catch this before.

All good! Thank you for asking @leog 🙌

alishaz-polymath committed:
fixing e2e tests

alishaz-polymath committed:
fix test --WIP

alishaz-polymath committed:
attempt fix test & type --WIP

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
fix duplicate locale en entry

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
fix locked state persistence

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
adds private URL locked/unlocked functionality

alishaz-polymath committed:
fix tests

alishaz-polymath committed:
Fix issue where locked labels showed up in non-managed events

alishaz-polymath committed:
adds e2e test step

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
update new test

alishaz-polymath reviewed (commented):
Most things are self-explanatory, comments added to clarify the potentially confusing bits. Happy to provide more context, if needed.

alishaz-polymath committed:
minor fixes

alishaz-polymath committed:
fix test

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

joeauyeung commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -106,23 +106,24 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupProps,
+  const lockTimeZoneToggleOnBookingPageLocked = shouldLockDisableProps("lockTimeZoneToggleOnBookingPage");
 
   const closeEventNameTip = () => setShowEventNameTip(false);
   const displayDestinationCalendarSelector =
     !!connectedCalendarsQuery.data?.connectedCalendars.length && !team;
```
Would this also be a good opprotunity to allow children event types to select their own destination calendar?

alishaz-polymath replied:
Yeah, good spot. I think this makes sense!

joeauyeung reviewed (dismissed):
Great PR @alishaz-polymath I've gone through and tested each fields except for "Apps" and "Workflows" since I think those would need a bit more time. I've left a few comments so far.

### Blocking
- When an admin locks a previously unlocked field, I think we should push the parent field down to the children. Ex. if a child event type has a different title from the parent, if a field is going to be locked again there isn't a way to change the title of the child unless the field is unlocked, changed back, and then locked again
- I noticed when the parent makes changes and saves it's propagated to children event types even if that field is unlocked and there's a different value set from the parent
- When multiple durations are set, the toggle to lock/unlock the default value isn't changing
![image](https://github.com/calcom/cal.com/assets/65426560/805e8ef9-d3bb-46b9-9d06-4092e61823e0)
- Member's default location isn't being set in the field unless we choose something else
![image](https://github.com/calcom/cal.com/assets/65426560/5248f5e7-4452-4643-963b-1fbacc777ee7)
- When selecting offset start times, there's two toggles for lock/unlock but they're tied together. I noticed other options that are similar to this don't have the same toggles for the sub options. I think we should decide if we want to offer this level of granularity or not
![image](https://github.com/calcom/cal.com/assets/65426560/c9901195-e7c5-4ce8-9d07-90d88aecd9e4)
- When toggling the "Enable Private URL", on and off we'll sometimes run into the prisma error `An operation failed because it depends on one or more records that were required but not found. No 'HashedLink' record was found for a nested delete on relation 'EventTypeToHashedLink'`. The second time I hit the save button it goes through.


### Non-blocking
- I noticed that for the single line text fields, when toggling the switch it focuses on the text field. I think we can make this looked more polished if we had these separated
- Likewise I'm not a huge fan of the changing sizes of the field lock toggle. Could it be a set width? 
![V2 managed field](https://github.com/calcom/cal.com/assets/65426560/1f9142d3-8a0b-469a-8b06-035925e5c301)
- Should we also lock/unlock the number of seats offered?
![image](https://github.com/calcom/cal.com/assets/65426560/d004dd8b-5d89-466c-97e5-54416421f2db)

alishaz-polymath committed:
address changes request Part 1

fixed width on the lock badge
members default location as default selection
bugfix for offset toggle duplicate within itself

alishaz-polymath commented:
> Great PR @alishaz-polymath I've gone through and tested each fields except for "Apps" and "Workflows" since I think those would need a bit more time. I've left a few comments so far.
> ### Blocking
> 
>     * When an admin locks a previously unlocked field, I think we should push the parent field down to the children. Ex. if a child event type has a different title from the parent, if a field is going to be locked again there isn't a way to change the title of the child unless the field is unlocked, changed back, and then locked again
  Asked @ciaranha for his inputs here ^, once we have finalized the ideal flow, I'll implement the changes
_________________________________________________________________________________
>     * I noticed when the parent makes changes and saves it's propagated to children event types even if that field is unlocked and there's a different value set from the parent
> 

Asked @ciaranha for his inputs here ^, once we have finalized the ideal flow, I'll implement the changes
_________________________________________________________________________________
>     * When multiple durations are set, the toggle to lock/unlock the default value isn't changing

Resolved ✅ 
_________________________________________________________________________________
>     * Member's default location isn't being set in the field unless we choose something else

Resolved ✅ 
 _________________________________________________________________________________
>     * When selecting offset start times, there's two toggles for lock/unlock but they're tied together. I noticed other options that are similar to this don't have the same toggles for the sub options. I think we should decide if we want to offer this level of granularity or not

This was unintended, fixed and Resolved ✅ 
_________________________________________________________________________________
>     * When toggling the "Enable Private URL", on and off we'll sometimes run into the prisma error `An operation failed because it depends on one or more records that were required but not found. No 'HashedLink' record was found for a nested delete on relation 'EventTypeToHashedLink'`. The second time I hit the save button it goes through.
 
I couldn't re-create this at all on my end, so perhaps I'd need more information on how to recreate this.
_________________________________________________________________________________
> 
> ### Non-blocking
> 
>     * I noticed that for the single line text fields, when toggling the switch it focuses on the text field. I think we can make this looked more polished if we had these separated
> 

Not entirely sure I follow here, when I click on the toggle, it doesn't switch focus to the text fields for me. Tested on Title
_________________________________________________________________________________
>     * Likewise I'm not a huge fan of the changing sizes of the field lock toggle. Could it be a set width?

I agree, Done & Resolved ✅ 
_________________________________________________________________________________
>     * Should we also lock/unlock the number of seats offered?

I think that's overkill tbh. Just the parent field being locked/unlocked should be enough IMO.

alishaz-polymath committed:
fixes locked/unlocked label for child event types

alishaz-polymath committed:
enable choice of destination calendar for children

CarinaWolli commented on packages/features/ee/workflows/components/EventWorkflowsTab.tsx:
```diff
@@ -223,13 +232,17 @@ function EventWorkflowsTab(props: Props) {
             .includes(workflow.id)
       );
-      setSortedWorkflows(activeWorkflows.concat(disabledWorkflows));
+      const allSortedWorkflows = workflowsDisableProps.isLocked
```
for the parent managed event type we want to list all team workflows here even when locked

CarinaWolli commented on apps/web/components/eventtype/EventAppsTab.tsx:
```diff
@@ -133,12 +136,28 @@ export const EventAppsTab = ({ eventType }: { eventType: EventType }) => {
+                  {lockedText.slice(1)} {isManagedEventType ? "for members" : "by team admins"}
+                </Trans>
+              }
+              actions={<div className="flex h-full items-center">{appsDisableProps.LockedIcon}</div>}
```
If enabling/disabling one toggle on this page all toggles changes
<img width="999" alt="Screenshot 2024-03-01 at 3 31 50 PM" src="https://github.com/calcom/cal.com/assets/30310907/53faad89-467f-4a5c-81b4-3bc1c8a944ff">

alishaz-polymath replied:
This is too granular for now, so only added the locked/unlocked for All apps.

CarinaWolli reviewed (dismissed):
Awesome work 👏🏻 I tested everything and found a few small things that need to be addressed 🙏

alishaz-polymath committed:
Fixes width for simple lock badge

joeauyeung commented:
@alishaz-polymath added some clarifying points

- When toggling the "Enable Private URL", on and off we'll sometimes run into the prisma error `An operation failed because it depends on one or more records that were required but not found. No 'HashedLink' record was found for a nested delete on relation 'EventTypeToHashedLink'`. The second time I hit the save button it goes through.
	- This might be related to propagating changes from the parent down to the child. Since the private URL is regenerated when the option is toggled, when the parent toggles the option the child private URL becomes broken. Which is causing the URL to become invalid or causes the DB issue above

- I noticed that for the single line text fields, when toggling the switch it focuses on the text field. I think we can make this looked more polished if we had these separated
	- In that GIF when we click the toggle, the field turns white and we can start typing. For other fields we don't do that so I think we should keep it consistent

joeauyeung committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
fix type

alishaz-polymath committed:
fixes workflows list and apps for managed event type

alishaz-polymath committed:
restricts creation of managed event types to EE only

alishaz-polymath commented on apps/web/components/eventtype/EventWebhooksTab.tsx:
```diff
@@ -4,8 +4,10 @@ import { Trans } from "next-i18next";
 import Link from "next/link";
```
We can line it up for a follow up

alishaz-polymath committed:
further fixes --WIP

alishaz-polymath committed:
fix unit test for handleChildrenEventTypes

alishaz-polymath committed:
fix test --WIP

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
fix type err --WIP

alishaz-polymath committed:
fix type err in test

alishaz-polymath committed:
fix childevent payload

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

CarinaWolli commented:
Is this ready for review again or still something missing?

alishaz-polymath commented:
> Is this ready for review again or still something missing?

Just waiting for the checks to pass 😅

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

joeauyeung committed:
Merge branch 'main' into feat/v2-managed-events

joeauyeung reviewed (dismissed):
Retested my review comments
## Passing
- When an admin locks a previously unlocked field, I think we should push the parent field down to the children ✅ 

## Blocking
- Still getting errors around the private URL field. When the field is unlocked and the child has the option enabled. When toggling the option on the parent I think it's trying to generate new URLs for the children. If it's using Prisma's `createMany` I would look at the `skipDuplicates` option.
![CleanShot 2024-03-08 at 11 39 31@2x](https://github.com/calcom/cal.com/assets/65426560/633ed23d-69d5-4eec-8193-66cbe6c26e75)

alishaz-polymath commented:
> Retested my review comments
> ## Passing
> 
>     * When an admin locks a previously unlocked field, I think we should push the parent field down to the children ✅
> 
> 
> ## Blocking
> 
>     * Still getting errors around the private URL field. When the field is unlocked and the child has the option enabled. When toggling the option on the parent I think it's trying to generate new URLs for the children. If it's using Prisma's `createMany` I would look at the `skipDuplicates` option.
>       ![CleanShot 2024-03-08 at 11 39 31@2x](https://private-user-images.githubusercontent.com/65426560/311301161-633ed23d-69d5-4eec-8193-66cbe6c26e75.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk5MTY5NzEsIm5iZiI6MTcwOTkxNjY3MSwicGF0aCI6Ii82NTQyNjU2MC8zMTEzMDExNjEtNjMzZWQyM2QtNjlkNS00ZWVjLTgxOTMtNjZjYmU2YzI2ZTc1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA4VDE2NTExMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThkM2E3ZmM5ZGY4NzVkMzk5NGZlNWY1NjM5Y2RlM2ExZGNjZjA4NzU3YzJiNTE3YjE4YjljN2ZhMWE2NDhmNmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mpdCEGiSVq_obYqQtPp8-3Wt8Ib24wt6-GgskYoRJ1w)

Just managed to re-create the issue. Working on a fix 🙏

alishaz-polymath committed:
fixes hashedLink bug

alishaz-polymath committed:
fix test step title

alishaz-polymath committed:
Merge branch 'feat/v2-managed-events' of https://github.com/calcom/cal.com into feat/v2-managed-events

CarinaWolli reviewed (commented):
Now a user can enable a workflow on a managed event type even when it is locked. All my other comments are fixed 🙏

joeauyeung reviewed (commented):
All my blocking comments have been resolved. Great work @alishaz-polymath! I'll wait until @CarinaWolli's comments have been addressed before approving.

alishaz-polymath committed:
lock workflow create button when workflow is locked

alishaz-polymath committed:
Merge branch 'main' into feat/v2-managed-events

alishaz-polymath commented on packages/features/ee/workflows/components/WorkflowDetailsPage.tsx:
```diff
@@ -54,12 +54,18 @@ export default function WorkflowDetailsPage(props: Props) {
-              eventType.children && eventType.children.length ? `(+${eventType.children.length})` : ``
-            }`,
-          })),
+          ...group.eventTypes
```
In this case as well, we do not want to allow users to create workflow when workflows are locked, correct? That should only be doable on Managed Event Level and not Child Event level, unless workflows are unlocked?

CarinaWolli replied:
Yes. When locked, the parent managed event type can still set the event type active on a team workflow. But the user can't set a locked child managed event active on one of their workflows

alishaz-polymath replied:
I think the latest changes sorts this. Can you test and confirm unless I missed something? 🙏

alishaz-polymath committed:
fix workflow detail

CarinaWolli reviewed (dismissed):
> All my blocking comments have been resolved. Great work @alishaz-polymath! I'll wait until @CarinaWolli's comments have been addressed before approving.

All my comments are also fixed. Looking good now 🙌🏻

alishaz-polymath commented:
Since this is Core, let's wait for Foundations team to give a green signal before we merge this 🙏

emrysal committed:
Don't rely on parameter ordering, use object instead

emrysal committed:
Removed console.log

emrysal commented on packages/features/ee/managed-event-types/hooks/useLockedFieldsManager.tsx:
```diff
@@ -1,27 +1,102 @@
 // eslint-disable-next-line no-restricted-imports
```
Just feedback on the naming, as this file doesn't only include the hook file this may be confusing.

alishaz-polymath replied:
Oops, a remnant from earlier changes 🤦 
Nice catch

emrysal commented on packages/ui/components/form/switch/SettingsToggle.tsx:
```diff
@@ -55,7 +55,8 @@ function SettingsToggle({
                   <Label
-                    className={classNames("mt-0.5 text-base font-semibold leading-none", labelClassName)}>
+                    className={classNames("mt-0.5 text-base font-semibold leading-none", labelClassName)}
+                    htmlFor="">
```
❓ I assume this was discussed elsewhere 😅  - OK

emrysal reviewed (dismissed):
Pushed some changes, but overall really happy with how this works. Happy to approve 👍  Well done @alishaz-polymath

emrysal committed:
Missed one

emrysal committed:
Merge branch 'main' into feat/v2-managed-events
